src/Avalonia.Base/Input/AccessKeyHandler.cs

```diff
@@ -255,5 +280,298 @@ private void MainMenuClosed(object? sender, EventArgs e)
+        /// <param name="key"></param>
+        /// <returns>AccessKeyInformation with target for the access key.</returns>
+        private static AccessKeyInformation GetTargetForElement(IInputElement? element, string key)
+        {
+            var info = new AccessKeyInformation();
+            if (element == null)
+                return info;
+
+            var args = new AccessKeyPressedEventArgs(key);
+            element.RaiseEvent(args);
+            info.Target = args.Target;
+
+            return info;
+        }
+
+        /// <summary>
+        /// Checks if the focused element is a descendent of the owner.
+        /// </summary>
+        /// <param name="owner">The owner to check.</param>
+        /// <returns>If focused element is decendant of owner <c>true</c>, otherwise <c>false</c>. </returns>
+        private static bool IsFocusWithinOwner(IInputRoot owner)
+        {
+            var focusedElement = KeyboardDevice.Instance?.FocusedElement;
+            if (focusedElement is not InputElement inputElement)
+                return false;
+
+            var isAncestorOf = owner is Visual root && root.IsVisualAncestorOf(inputElement);
+            return isAncestorOf;
+        }
+
+        /// <summary>
+        /// Sorts the list of targets according to logical ancestors in the hierarchy
+        /// so that child elements, for example within in the content of a tab,
+        /// are processed before the next parent item i.e. the next tab item.
+        /// </summary>
+        private static List<IInputElement> SortByHierarchy(List<IInputElement> targets)
+        {
+            var sorted = new List<IInputElement>(targets.Count);
+            var queue = new Queue<IInputElement>(targets);
+            while (queue.Count > 0)
+            {
+                var element = queue.Dequeue();
+            
+                // if the element was already added, do nothing
+                if (sorted.Contains(element))
+                    continue;
+            
+                // add the element itself
+                sorted.Add(element);
+
+                // if the element is not a potential parent, do nothing
+                if (element is not ILogical parentElement) 
+                    continue;
+                
+                // add all descendants of the element
+                sorted.AddRange(queue
+                    .Where(child => parentElement
+                        .IsLogicalAncestorOf(child as ILogical)));
+            }
+            return sorted;
+        }
+
+        private enum ProcessKeyResult
+        {
+            NoMatch,
+            MoreMatches,
+            LastMatch
+        }
+
+        private struct AccessKeyInformation
+        {
+            public IInputElement? Target { get; set; }
+        }
+    }
+
+    /// <summary>
+    /// The inputs to an AccessKeyPressedEventHandler
+    /// </summary>
+    public class AccessKeyPressedEventArgs : RoutedEventArgs
```

Comment:
I've missed this previously, but please keep the event arguments `internal` for now, since the event itself is internal.

We should definitely expose the events publicly at some point to allow users to participate in access key handling, but that should be another PR with some API review. I don't want to block this PR because of that :)